tools/tpm2_pcr_extend.c

```diff
@@ -0,0 +1,151 @@
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <ctype.h>
+#include <getopt.h>
+
+#include <sapi/tpm20.h>
+
+#include "files.h"
+#include "log.h"
+#include "main.h"
+#include "options.h"
+#include "tpm2_alg_util.h"
+#include "tpm2_util.h"
+
+typedef struct tpm_pcr_extend_ctx tpm_pcr_extend_ctx;
+struct tpm_pcr_extend_ctx {
+    TPML_DIGEST_VALUES digests;
+    TPMI_DH_PCR pcr_index;
+    TSS2_SYS_CONTEXT *sapi_context;
+};
+
+static bool init(int argc, char *argv[], tpm_pcr_extend_ctx *ctx) {
+
+    static struct option long_options[] = {
+        { "digest-list", required_argument, NULL, 'D'},
+        { "pcr-index",   required_argument, NULL, 'i'},
+        { NULL,          no_argument,       NULL, '\0'}
+    };
+
+    int opt;
+    while ((opt = getopt_long(argc, argv, "D:i:", long_options, NULL))
+            != -1) {
+        switch (opt) {
+        case 'D':
+            if (!pcr_parse_digest_list(optarg, &ctx->digests)) {
+                LOG_ERR("Could not parse digest-list, got: \"%s\"", optarg);
+                return false;
+            }
+            break;
+        case 'i':
+            if (!tpm2_util_string_to_uint32(optarg, &ctx->pcr_index)) {
+                LOG_ERR("Could not convert to pcr index, got: \"%s\"!",
+                        optarg);
+                return false;
+            }
+            break;
+        case ':':
+            LOG_ERR("Argument %c needs a value!\n", optopt);
+            return false;
+        case '?':
+            LOG_ERR("Unknown Argument: %c\n", optopt);
+            return false;
+        default:
+            LOG_ERR("?? getopt returned character code 0%o ??\n", opt);
+            return false;
+        }
+    }
+
+    if (!ctx->digests.count) {
```

Comment:
I think I am going to move the digest option to argv.

Would we really need -f if we can specify input file in the digest specification directly?

tpm2_pcr_extend -D sha1:/foo/bar